### ISSUE TITLE
onNext not working as expected. Console logs not even running.

### ISSUE THREAD

**Issue by alexanderbanks** [linkedin/hopscotch#298]
My requirement is to show a hopscotch popup on some dynamic content. The first is an element that there could be many of, so it doesn't normally have an ID.  My original idea was something like:

```
         {
            title: "Your Name",
            content: "This is your name",
            target: "profileName",
            placement: "bottom",
            onNext: function() {
               $('.some-class').first().attr('id', 'tourItem');
            }
          },
          {
             title: 'Some Title',
             content: 'Some Content',
             target: 'tourItem'
             placement: 'bottom'
          }
```

But this approach did not work. onNext doesn't appear to modify the element at all? Also, if I add a console.log into the onNext callback, it never shows.

I can provide more source if necessary, as it seems I may be fundamentally misunderstanding how onNext is supposed to work? I'm also facing this issue while trying to show a popup on an element that is defaultly hidden.

**Comment by travstone** [comment]
Not sure what you mean exactly, but onNext will only fire when the 'Next' button is clicked. If you want to do something dynamic in your first step, you might want to define an 'onShow' method for the step, which will fire as soon as the bubble becomes visible.

For the second topic, if the 'hidden' element is actually 'display: none', then an offset relative to that element probably cannot be calculated. I'm not sure how hopscotch handles 'visibility: hidden'.

**Comment by alexanderbanks** [comment]
@travstone Yes, so my understanding is in the above code, when the 'Next' button is clicked, before the new popup is rendered the DOM will be modified to set the ID attribute to 'tourItem'. I am saying that it does not appear the onNext function is ever running.

**Comment by travstone** [comment]
Hmm- do you have a Tour-level 'onNext' defined? I seem to recall that this may override a step-level 'onNext'. At any rate, onNext does fire in a simple case: http://codepen.io/trav_stone/pen/ZOVyap, so there must be some other important detail here.

Also, I am not positive that onNext gets called _before_ the new bubble is rendered- you may want to trace that in the console.

**Comment by alexanderbanks** [comment]
@travstone So I guess I'm confused. In the first onNext I'm adding the ID to the element, but I need that to happen before the second step tries to query for the ID. Is there any sort of guaranteed way to run code X before code Y in Hopscotch? Also thanks for the timely response.

**Comment by travstone** [comment]
Sure... happy to help if I can!

AFAIK, the only way to do what you're interested in is to make an onShow property for the first step (in your example). I had a similar need a while back, and that's what I did. You have to go out of your way (intentionally or in error) to make Hopscotch *not* execute the steps in (array) order, so if you're using it in the standard way one could say that step 1 is guaranteed to execute before step 2.

**Comment by zimmi88** [comment]
Yeah, unfortunately, `nextStep` doesn't get called until the tour has determined what step it's going to next. In the event there's not another step that matches, the tour will end and `onNext` won't get called; if there is a match, Hopscotch will (usually) skip to that step. IMO, this is counterintuitive and should be fixed (you'd expect `onNext` to fire before Hopscotch tries to move on), but for now I would recommend using `onShow` of the previous step as a workaround, or even better (if possible), do any DOM manipulation to add IDs before starting the tour.

Another option... `target` can be a query selector. If you have `jQuery` on the page, Hopscotch will use that (at least in the current version), otherwise Hopscotch will use `document.querySelector()`. In both cases, the first match is used, should there be multiple matches.

**Comment by bharatpec** [comment]
I faced the same issue. onNext wasn't being called. I observed that the next step 'target' was an Id of some element which was not present at the previous state. The tour couldn't find the element and hence, that tour ended there and then. Just make sure the target id of next element is accessible on the previous step.
